pkg/ottl/ottlfuncs/func_scale.go

```diff
@@ -0,0 +1,137 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ottlfuncs // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+type ScaleArguments[K any] struct {
+	Value      ottl.Getter[K]
+	Multiplier float64
+}
+
+func NewScaleFactory[K any]() ottl.Factory[K] {
+	return ottl.NewFactory("Scale", &ScaleArguments[K]{}, createScaleFunction[K])
+}
+
+func createScaleFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[K], error) {
+	args, ok := oArgs.(*ScaleArguments[K])
+
+	if !ok {
+		return nil, fmt.Errorf("ScaleFactory args must be of type *ScaleArguments[K]")
+	}
+
+	return Scale(args.Value, args.Multiplier)
+}
+
+func Scale[K any](getter ottl.Getter[K], multiplier float64) (ottl.ExprFunc[K], error) {
+	return func(ctx context.Context, tCtx K) (any, error) {
+		got, err := getter.Get(ctx, tCtx)
+		if err != nil {
+			return nil, err
+		}
+
+		switch value := got.(type) {
+		case float64:
+			return value * multiplier, nil
+		case int64:
+			return float64(value) * multiplier, nil
+		case pmetric.NumberDataPointSlice:
+			scaledMetric := pmetric.NewNumberDataPointSlice()
+			value.CopyTo(scaledMetric)
+			scaleMetric(scaledMetric, multiplier)
+			return scaledMetric, nil
+		case pmetric.HistogramDataPointSlice:
+			scaledMetric := pmetric.NewHistogramDataPointSlice()
+			value.CopyTo(scaledMetric)
+			scaleHistogram(scaledMetric, multiplier)
+			return scaledMetric, nil
+		case pmetric.SummaryDataPointValueAtQuantileSlice:
+			scaledMetric := pmetric.NewSummaryDataPointValueAtQuantileSlice()
+			value.CopyTo(scaledMetric)
+			scaleSummaryDataPointValueAtQuantileSlice(scaledMetric, multiplier)
+			return scaledMetric, nil
+		case pmetric.ExemplarSlice:
+			scaledSlice := pmetric.NewExemplarSlice()
+			value.CopyTo(scaledSlice)
+			scaleExemplarSlice(scaledSlice, multiplier)
+			return scaledSlice, nil
+		case pmetric.ExponentialHistogramDataPointSlice:
+			return nil, errors.New("exponential histograms are not supported by the 'Scale' function")
+		default:
+			return nil, errors.New("unsupported data type")
```

Comment:
```suggestion
			return nil, fmt.Errorf("unsupported data type: '%T'", val)
```
We should also call this out here, it would be useful for debugging.